### PR TITLE
Add Featureform version to the CLI and Dashhoard

### PR DESCRIPTION
**Pull request by RedLeader16**
# Description

This PR adds functionality to display the client and cluster version in the `cli`, and the cluster version in the `dashboard` application. It also includes unit tests for the backend and frontend. 

The `--local` flag is required when running in local mode.

## Type of change
`CLI`
Client Package Version and Cluster Version:
Screenshot 2023-05-26 at 2 48 27 PM

`local mode and hosted mode`
version and version local 2

`endpoint`
The metadata-dashboard `/data/version` endpoint now returns a `versionMap`. Currently the map returns a single `version` property, and is expected to be expanded in the future. (The difference in values is just the configuration differences between local and k8s)
Screenshot 2023-05-22 at 5 24 13 PM

image

`docker`
Running in docker mode reads from the `FEATUREFORM_VERSION` env. variable. 
dash version docker
docker env variable

### Does this correspond to an open issue?

#812 

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have fixed any merge conflicts

### CODE REVIEW

File: charts/featureform/charts/dashboard-metadata/values.yaml [hunk 1]
```diff
@@ -15,6 +15,9 @@ imagePullSecrets: []
 nameOverride: ""
 fullnameOverride: ""
 
+#current application version
+featureformVersion: "v0.8.1"
```
Comment (RedLeader16) [inline charts/featureform/charts/dashboard-metadata/values.yaml:19]: thoughts on using the `values.yaml` file as the version prop home? pretty standard but we can store it elsewhere too.

**Comment by RedLeader16** [comment]
Should we drop the "Version:" in the text "Version: v0.8.1"? I think the "v" indicates it enough.

File: metadata/dashboard/dashboard_metadata.go [hunk 2]
```diff
@@ -1031,13 +1031,20 @@ func (m *MetadataServer) GetSearch(c *gin.Context) {
 	c.JSON(200, result)
 }
 
+func (m *MetadataServer) GetVersionMap(c *gin.Context) {
+	versionMap := map[string]string{}
+	versionMap["version"] = help.GetEnv("FEATUREFORM_VERSION", "Default")
```
[the remark this answers is not in the document]
Comment (RedLeader16) [reply metadata/dashboard/dashboard_metadata.go:1036]: Let's go with blank.

[the remark this answers is not in the document]
Comment (RedLeader16) [reply metadata/dashboard/dashboard_metadata.go:1036]: So something to the effect of, 

If a version is missing: We display nothing on the page. The versionMap returns `blank`. 
If there is a version: We display the version value only. ex. "v0.8.1". And not include the "Version:" prefix.

File: charts/featureform/charts/dashboard-metadata/values.yaml (hunk 1 above)
Comment (RedLeader16) [reply charts/featureform/charts/dashboard-metadata/values.yaml:19]: updating from our side chat: 

we'll use the `.Values.global.version | default .Chart.AppVersion` approach. that way we can just set the global as needed during upgrades etc.

File: metadata/dashboard/dashboard_metadata.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (RedLeader16) [reply metadata/dashboard/dashboard_metadata.go]: decided to keep the "Version:" prefix.

File: charts/featureform/charts/dashboard-metadata/values.yaml (hunk 1 above)
Comment (RedLeader16) [reply charts/featureform/charts/dashboard-metadata/values.yaml]: and done.

File: metadata/dashboard/dashboard_metadata.go [hunk 3]
```diff
@@ -1031,13 +1031,20 @@ func (m *MetadataServer) GetSearch(c *gin.Context) {
 	c.JSON(200, result)
 }
 
+func (m *MetadataServer) GetVersionMap(c *gin.Context) {
+	versionMap := map[string]string{}
```
[the remark this answers is not in the document]
Comment (RedLeader16) [reply metadata/dashboard/dashboard_metadata.go]: done.

File: client/src/featureform/cli.py [hunk 4]
```diff
@@ -148,6 +149,11 @@ def list(host, cert, insecure, local, resource_type):
 app.register_blueprint(dashboard_app)
 
 
+@cli.command()
+def version():
+    print("Client Version: {}".format(get_package_version()))
```
Comment (sdreyer) [inline client/src/featureform/cli.py:154]: Would we be able to do a GET here and if FEATUREFORM_HOST env var is set then get the cluster version?
Then have:
```
Client Version:
Cluster Version:
```
```
If FEATUREFORM_HOST != ""
try:
    version = Get <host>/data/version
except:
    version = "Could not fetch version"

```

Comment (RedLeader16) [reply client/src/featureform/cli.py:154]: Yeah that should be possible, I'll give it a shot and report back.

Comment (RedLeader16) [reply client/src/featureform/cli.py]: and done. I'll get with you first to double check some assumptions. 

Screenshot 2023-05-26 at 2 48 27 PM

Comment (RedLeader16) [reply client/src/featureform/cli.py]: and done. I'll get with you first to double check some assumptions.
image

File: client/src/featureform/tls.py [hunk 5]
```diff
@@ -15,3 +20,26 @@ def secure_channel(host, cert_path):
...
+def fetch_cluster_version(version_url=""):
+    requests.packages.urllib3.disable_warnings()
+    res = requests.get(url=version_url, verify=False)
+    response = json.loads(res.text)
+    return response["version"]
+
+
+def get_version_local():
+    local_port = os.getenv("LOCALMODE_DASHBOARD_PORT", 3000)
+    version_url = "localhost:{}/data/version".format(local_port)
```
Comment (ahmadnazeri) [inline client/src/featureform/tls.py:34]: I would recommend using `f-strings`, just more readable. 
```
version_url = f"localhost:{local_port}/data/version"

return fetch_cluster_version(f"{insecure_protocol}{version_url}")
```

**Review by ahmadnazeri — APPROVED** [review]
lgtm; just a small recommendation of using f-strings on the python side

File: client/src/featureform/cli.py [hunk 6]
```diff
@@ -148,6 +150,32 @@ def list(host, cert, insecure, local, resource_type):
...
+            cluster_version = "No host is set."
+    except:
+        cluster_version = "Cannot retrieve: Check your FEATUREFORM_HOST value. If using local mode, use the --local flag."
+
+    print(
+        """
+    Client Version: {}
+    Cluster Version: {}
+    """.format(
+            client_cersion, cluster_version
```
Comment (ahmadnazeri) [inline client/src/featureform/cli.py:174]: is this suppose to be `client_version`?

File: client/src/featureform/cli.py [hunk 7]
```diff
@@ -148,6 +150,32 @@ def list(host, cert, insecure, local, resource_type):
 app.register_blueprint(dashboard_app)
 
 
+@cli.command()
+@click.option("--local", is_flag=True, help="Required for local mode only")
+def version(local):
+    client_cersion = get_package_version()
```
Comment (sdreyer) [inline client/src/featureform/cli.py:156]: typo

**Review by sdreyer — DISMISSED** [review]

File: client/src/featureform/tls.py (hunk 5 above)
Comment (RedLeader16) [reply client/src/featureform/tls.py:34]: agreed. changing.

File: client/src/featureform/cli.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (RedLeader16) [reply client/src/featureform/cli.py]: yup changed.

Comment (RedLeader16) [reply client/src/featureform/cli.py]: done

File: client/src/featureform/dashboard_metadata.py [hunk 8]
```diff
@@ -62,6 +62,11 @@ def deliver_static(asset):
     return dashboard_app.send_static_file("static/" + asset)
 
 
+@dashboard_app.route("/data/version")
+def version():
+    return {"version": get_package_version()}
```
Comment (ahmadnazeri) [inline client/src/featureform/dashboard_metadata.py:67]: Is this the same as the client version? do we want to know the cluster version?

Comment (RedLeader16) [reply client/src/featureform/dashboard_metadata.py:67]: IIRC we're using the python package version in local mode.

**Review by sdreyer — APPROVED** [review]

**Pull request merged**